File: oxidize-golang/core/mesh/topology_extended.go
package mesh

import (
	"sort"
	"sync"
	"time"
)

// NodeCapabilities mirrors NodeCapabilities advertised during discovery.
type NodeCapabilities struct {
	DeviceType  string
	MemoryBytes uint64
	CPUThreads  int
	CanShard    bool
	Tags        map[string]string
}

// AggregateCapabilities mirrors AggregateCapabilities: a summary across the
// whole mesh.
type AggregateCapabilities struct {
	NodeCount        int
	TotalMemoryBytes uint64
	TotalCPUThreads  int
	CanShardNodes    int
	DeviceTypes      []string
}

// CapTopologyNode mirrors TopologyNode: a node tracked with capabilities and
// liveness metadata used for eviction and election tie-breaking.
type CapTopologyNode struct {
	PeerID       string
	Capabilities NodeCapabilities
	CommandsSeen uint64
	Seniority    uint64
	LastSeen     time.Time
	JoinedAt     time.Time
}

// IsStale reports whether the node has not been seen within timeout.
func (n *CapTopologyNode) IsStale(timeout time.Duration) bool {
	if n.LastSeen.IsZero() {
		return true
	}
	return time.Since(n.LastSeen) > timeout
}

// CapTopologyGraph mirrors the capability-aware TopologyGraph. It is additive to
// the existing TopologyGraph (which tracks plain MeshNode adjacency) and tracks
// per-peer NodeCapabilities for aggregation and stale eviction.
type CapTopologyGraph struct {
	mu          sync.Mutex
	nodes       map[string]*CapTopologyNode
	LocalPeerID string
}

// NewCapTopologyGraph constructs an empty capability graph.
func NewCapTopologyGraph() *CapTopologyGraph {
	return &CapTopologyGraph{nodes: map[string]*CapTopologyNode{}}
}

// AddOrUpdateNode registers a peer or refreshes its capabilities and heartbeat.
func (g *CapTopologyGraph) AddOrUpdateNode(peerID string, caps NodeCapabilities) {
	g.mu.Lock()
	defer g.mu.Unlock()
	now := time.Now()
	if existing, ok := g.nodes[peerID]; ok {
		existing.Capabilities = caps
		existing.LastSeen = now
		return
	}
	g.nodes[peerID] = &CapTopologyNode{
		PeerID:       peerID,
		Capabilities: caps,
		LastSeen:     now,
		JoinedAt:     now,
	}
}

// RemoveNode drops a peer.
func (g *CapTopologyGraph) RemoveNode(peerID string) {
	g.mu.Lock()
	delete(g.nodes, peerID)
	g.mu.Unlock()
}

// Heartbeat refreshes the last-seen timestamp for a peer.
func (g *CapTopologyGraph) Heartbeat(peerID string) {
	g.mu.Lock()
	if n, ok := g.nodes[peerID]; ok {
		n.LastSeen = time.Now()
	}
	g.mu.Unlock()
}

// EvictStale removes nodes not seen within timeout and returns their IDs sorted.
// Mirrors evict_stale.
func (g *CapTopologyGraph) EvictStale(timeout time.Duration) []string {
	g.mu.Lock()
	defer g.mu.Unlock()
	var stale []string
	for id, n := range g.nodes {
		if n.IsStale(timeout) {
			stale = append(stale, id)
		}
	}
	for _, id := range stale {
		delete(g.nodes, id)
	}
	sort.Strings(stale)
	return stale
}

// PeerCount returns the number of known peers.
func (g *CapTopologyGraph) PeerCount() int {
	g.mu.Lock()
	defer g.mu.Unlock()
	return len(g.nodes)
}

// PeerIDs returns all known peer IDs (excluding the local peer if set), sorted.
func (g *CapTopologyGraph) PeerIDs() []string {
	g.mu.Lock()
	defer g.mu.Unlock()
	out := make([]string, 0, len(g.nodes))
	for id := range g.nodes {
		if g.LocalPeerID != "" && id == g.LocalPeerID {
			continue
		}
		out = append(out, id)
	}
	sort.Strings(out)
	return out
}

// CapabilitiesOf returns a peer's capabilities, if known.
func (g *CapTopologyGraph) CapabilitiesOf(peerID string) (NodeCapabilities, bool) {
	g.mu.Lock()
	defer g.mu.Unlock()
	n, ok := g.nodes[peerID]
	if !ok {
		return NodeCapabilities{}, false
	}
	return n.Capabilities, true
}

// AggregateCapabilities summarizes capabilities across all peers. Mirrors
// aggregate_capabilities.
func (g *CapTopologyGraph) AggregateCapabilities() AggregateCapabilities {
	g.mu.Lock()
	defer g.mu.Unlock()
	agg := AggregateCapabilities{NodeCount: len(g.nodes)}
	seen := map[string]struct{}{}
	for _, n := range g.nodes {
		agg.TotalMemoryBytes += n.Capabilities.MemoryBytes
		agg.TotalCPUThreads += n.Capabilities.CPUThreads
		if n.Capabilities.CanShard {
			agg.CanShardNodes++
		}
		if _, ok := seen[n.Capabilities.DeviceType]; !ok {
			seen[n.Capabilities.DeviceType] = struct{}{}
			agg.DeviceTypes = append(agg.DeviceTypes, n.Capabilities.DeviceType)
		}
	}
	sort.Strings(agg.DeviceTypes)
	return agg
}

// ShardableNodes returns the IDs of peers that can act as shard workers,
// sorted deterministically. Implements capability filtering for can_shard.
func (g *CapTopologyGraph) ShardableNodes() []string {
	g.mu.Lock()
	defer g.mu.Unlock()
	var out []string
	for id, n := range g.nodes {
		if n.Capabilities.CanShard {
			out = append(out, id)
		}
	}
	sort.Strings(out)
	return out
}

// SetSeniority assigns a seniority score used for election tie-breaking.
func (g *CapTopologyGraph) SetSeniority(peerID string, seniority uint64) {
	g.mu.Lock()
	if n, ok := g.nodes[peerID]; ok {
		n.Seniority = seniority
	}
	g.mu.Unlock()
}

File: oxidize-python/oxidize_python/core/mesh/topology_extended.py
"""Capability-aware topology mirroring oxidize-golang/core/mesh/topology_extended.go.

Extends the simple ``TopologyGraph`` with per-peer capability tracking, stale
node eviction, capability aggregation and shardable-node filtering.
"""

from __future__ import annotations

import threading
import time
from dataclasses import dataclass, field


@dataclass
class NodeCapabilities:
    """Capabilities advertised during discovery (mirrors NodeCapabilities)."""

    device_type: str = ""
    memory_bytes: int = 0
    cpu_threads: int = 0
    can_shard: bool = False
    tags: dict[str, str] = field(default_factory=dict)


@dataclass
class AggregateCapabilities:
    """Mesh-wide capability summary (mirrors AggregateCapabilities)."""

    node_count: int = 0
    total_memory_bytes: int = 0
    total_cpu_threads: int = 0
    can_shard_nodes: int = 0
    device_types: list[str] = field(default_factory=list)


@dataclass
class CapTopologyNode:
    """Capability-tracked node with liveness metadata (mirrors TopologyNode)."""

    peer_id: str
    capabilities: NodeCapabilities
    commands_seen: int = 0
    seniority: int = 0
    last_seen: float = 0.0
    joined_at: float = 0.0

    def is_stale(self, timeout: float) -> bool:
        if self.last_seen <= 0.0:
            return True
        return (time.monotonic() - self.last_seen) > timeout


class CapTopologyGraph:
    """Capability-aware topology graph (mirrors CapTopologyGraph).

    Additive to the existing ``TopologyGraph``; tracks per-peer
    ``NodeCapabilities`` for aggregation and stale eviction.
    """

    def __init__(self, local_peer_id: str = "") -> None:
        self._mu = threading.Lock()
        self._nodes: dict[str, CapTopologyNode] = {}
        self.local_peer_id = local_peer_id

    def add_or_update_node(self, peer_id: str, caps: NodeCapabilities) -> None:
        with self._mu:
            now = time.monotonic()
            existing = self._nodes.get(peer_id)
            if existing is not None:
                existing.capabilities = caps
                existing.last_seen = now
                return
            self._nodes[peer_id] = CapTopologyNode(
                peer_id=peer_id,
                capabilities=caps,
                last_seen=now,
                joined_at=now,
            )

    def remove_node(self, peer_id: str) -> None:
        with self._mu:
            self._nodes.pop(peer_id, None)

    def heartbeat(self, peer_id: str) -> None:
        with self._mu:
            n = self._nodes.get(peer_id)
            if n is not None:
                n.last_seen = time.monotonic()

    def evict_stale(self, timeout: float) -> list[str]:
        """Remove nodes not seen within ``timeout`` and return their sorted IDs."""
        with self._mu:
            stale = [pid for pid, n in self._nodes.items() if n.is_stale(timeout)]
            for pid in stale:
                del self._nodes[pid]
        stale.sort()
        return stale

    def peer_count(self) -> int:
        with self._mu:
            return len(self._nodes)

    def peer_ids(self) -> list[str]:
        """Return all known peer IDs excluding the local peer, sorted."""
        with self._mu:
            out = [
                pid
                for pid in self._nodes
                if not (self.local_peer_id and pid == self.local_peer_id)
            ]
        out.sort()
        return out

    def capabilities_of(self, peer_id: str) -> tuple[NodeCapabilities | None, bool]:
        with self._mu:
            n = self._nodes.get(peer_id)
        if n is None:
            return None, False
        return n.capabilities, True

    def aggregate_capabilities(self) -> AggregateCapabilities:
        """Summarize capabilities across all peers (mirrors aggregate_capabilities)."""
        with self._mu:
            agg = AggregateCapabilities(node_count=len(self._nodes))
            seen: set[str] = set()
            for n in self._nodes.values():
                agg.total_memory_bytes += n.capabilities.memory_bytes
                agg.total_cpu_threads += n.capabilities.cpu_threads
                if n.capabilities.can_shard:
                    agg.can_shard_nodes += 1
                if n.capabilities.device_type not in seen:
                    seen.add(n.capabilities.device_type)
                    agg.device_types.append(n.capabilities.device_type)
        agg.device_types.sort()
        return agg

    def shardable_nodes(self) -> list[str]:
        """Return IDs of peers that can act as shard workers, sorted."""
        with self._mu:
            out = [pid for pid, n in self._nodes.items() if n.capabilities.can_shard]
        out.sort()
        return out

    def set_seniority(self, peer_id: str, seniority: int) -> None:
        with self._mu:
            n = self._nodes.get(peer_id)
            if n is not None:
                n.seniority = seniority

    # Internal accessor used by run_election_round (holds no lock; callers must).
    def _snapshot_nodes(self) -> dict[str, CapTopologyNode]:
        with self._mu:
            return dict(self._nodes)
